fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/query/plan/cascades/rules/ImplementRecursiveDfsUnionRule.java

```diff
@@ -0,0 +1,124 @@
+public class ImplementRecursiveDfsUnionRule extends ImplementationCascadesRule<RecursiveUnionExpression> {
+
+    @Nonnull
+    private static final BindingMatcher<RecordQueryPlan> initialInnerPlanMatcher = anyPlan();
+
+    @Nonnull
+    private static final BindingMatcher<TempTableInsertPlan> initialPlanMatcher = tempTableInsertPlan(initialInnerPlanMatcher);
+
+    @Nonnull
+    private static final BindingMatcher<PlanPartition> innerPlanPartitionMatcher = anyPlanPartition();
+
+    @Nonnull
+    private static final BindingMatcher<Collection<PlanPartition>> innerPlanPartitionsMatcher = rollUpPartitions(any(innerPlanPartitionMatcher));
+
+    @Nonnull
+    private static final BindingMatcher<Reference> innerReferenceMatcher = planPartitions(innerPlanPartitionsMatcher);
+
+    @Nonnull
+    private static final BindingMatcher<Quantifier.ForEach> innerQunMatcher = forEachQuantifierOverRef(innerReferenceMatcher);
+
+    @Nonnull
+    private static final BindingMatcher<Quantifier.ForEach> mapTempTableQunMatcher = forEachQuantifier(mapPlan(tempTableScanPlan()));
+
+    @Nonnull
+    private static final BindingMatcher<Quantifier.ForEach> tempTableQunMatcher = forEachQuantifier(tempTableScanPlan());
+
+    @Nonnull
+    private static final BindingMatcher<SelectExpression> recursiveInnerSelectMatcher = selectExpression(
+            exactlyInAnyOrder(innerQunMatcher, tempTableQunMatcher.or(mapTempTableQunMatcher)));
+
+    @Nonnull
+    private static final BindingMatcher<TempTableInsertExpression> recursivePlanMatcher = tempTableInsertExpression(
+            forEachQuantifierOverRef(exploratoryMember(recursiveInnerSelectMatcher)));
+
+    @Nonnull
+    private static final BindingMatcher<RecursiveUnionExpression> root = recursiveUnionExpression(forEachQuantifier(initialPlanMatcher),
+            forEachQuantifier(recursivePlanMatcher)).where(preOrderTraversalIsAllowed());
+
+    public ImplementRecursiveDfsUnionRule() {
+        super(root);
+    }
+
+    @Override
+    public void onMatch(@Nonnull final ImplementationCascadesRuleCall call) {
+        final var recursiveUnion = call.get(root);
+        final var rootAlias = recursiveUnion.getInitialStateQuantifier().getAlias();
+        final var recursiveAlias = recursiveUnion.getRecursiveStateQuantifier().getAlias();
+
+        final var initialInnerPlan = call.get(initialInnerPlanMatcher);
+        final var recursiveInnerSelect = call.get(recursiveInnerSelectMatcher);
+        final var innerPlanPartition = call.get(innerPlanPartitionMatcher);
+        final var innerRef = call.get(innerReferenceMatcher);
+        final var innerQun = call.get(innerQunMatcher);
+        final var priorValueCorrelation = (call.getBindings().containsKey(tempTableQunMatcher)
+                                          ? call.get(tempTableQunMatcher)
+                                          : call.get(mapTempTableQunMatcher)).getAlias();
+
+        final var rootPlanRef = call.memoizePlan(initialInnerPlan);
```

Comment:
This needs to be `memoizeMemberPlansFromOther` to avoid reexploration of that plan.